### PR TITLE
Add fake tensor support for sparse compressed tensors

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

Fixes https://github.com/pytorch/pytorch/issues/99404

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120588
* #120562

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip @eellison

### CODE REVIEW

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -280,13 +296,16 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
                 if t.is_sparse:
                     is_leaf = safe_is_leaf(t)
                     r = callback(
-                        lambda: torch.ops.aten._sparse_coo_tensor_with_dims(
+                        lambda: torch.ops.aten._sparse_coo_tensor_with_dims_and_tensors(
                             t.sparse_dim(),
                             t.dense_dim(),
                             t.shape,
+                            t._indices().to(device="meta"),
+                            t._values().to(device="meta"),
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:304]: This conversion to meta looks suspicious

File: torch/_subclasses/meta_utils.py [hunk 2]
```diff
@@ -295,13 +314,48 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
                     # which means that it will get caught by fake tensor mode.
                     # Ordinarily this would error, but there's some logic in
                     # fake tensor ensure this doesn't happen.
-                    r._coalesced_(t.is_coalesced())
+                    # Update: with the introduction of is_coalesced kw
+                    # argument to sparse coo tensor factory function,
+                    # there is no need to call r._coalesced_(t.is_coalesced()).
+                    # Is this note still relevant?
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:320]: It looks like you still query `is_coalesced` in some situations below, not sure if you wanted to modify those too.

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:304]: You should look at the NT fakeification code, my expectation is you have to recursively fakeify indices/values

also cc @jbschlosser

**Comment by ezyang** [comment]
This should be similar to NT fakeify, so I've added @jbschlosser and @soulitzer too.

**Comment by PeimingLiu** [comment]
Thanks! Looking forward to this!

**Comment by pearu** [comment]
This PR is abandoned in favor of https://github.com/pytorch/pytorch/pull/120920

**Comment by pearu** [comment]
Closing as https://github.com/pytorch/pytorch/pull/120920 has landed

**Comment by thiagocrepaldi** [comment]
@pearu would https://github.com/pytorch/pytorch/issues/123345 be a corner case for this pr?

**Pull request closed without merging**